### ISSUE TITLE
Telemetry v2 EnvoyFilter leads to 25% regression in Pilot CPU

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#18824]
![2019-09-11_14-48-39](https://user-images.githubusercontent.com/623453/68535987-095a0800-0300-11ea-9326-002faa1718b9.png)

When enabling telemetry v2 using EnvoyFilter, Pilot total CPU usages goes up by 25%.

This should be moved to native Pilot code where it will likely have a ~0% impact on performance.

**Comment by bianpengyuan** [comment]
Is this a general issue that Pilot is not performant enough on dealing with EnvoyFilter, or is it particularly bad with stats filter config somehow?

**Comment by howardjohn** [comment]
As far as I know there is nothing in particular about the stats filter that makes the performance worse, it's just a pretty non-performant part of pilot since it needs to do a proto merge, which is slow

**Comment by kyessenov** [comment]
There is no need to do proto merge with stats filter. It should be close to native code since it only adds stuff to the xDS.

**Comment by howardjohn** [comment]
@kyessenov how do you add an arbitrary proto to another without using proto merge? If we can do that, then sounds great, but I am not sure how we can?

**Comment by howardjohn** [comment]
I added https://github.com/istio/istio/pull/18836 to get a benchmark going for this code path, then will see if this can be optimized

**Comment by kyessenov** [comment]
@howardjohn What I mean is that telemetry v2 is strict addition of filters without modifying anything created by pilot. As long as we apply the insertion before things are serialized, there should be a minimal performance cost (e.g. finding a pointer in a proto struct and setting it).

**Comment by howardjohn** [comment]
There are three perf issues:
* We unmarshal the HCM
* We add in the new config - for some reason we do 2 clones here, probably both are not necessary
* We marshal the HCM again

So there is a pretty large architectural issue that by the time we patch the config we have already serialized the HCM so we have to do a whole round trip again

**Comment by kyessenov** [comment]
Some performance improvements are blocked by conservative golang/protobuf code generation. So we'll have to wait until after v2 library ships which is probably many months in the future. I think we need to qualify our expectations. We should expect performance regression, but it's not clear whether that matters.

**Comment by howardjohn** [comment]
Numbers were a bit off from looking at just the profile, its actually using 1.6x more CPU

![2019-23-12_14-11-02](https://user-images.githubusercontent.com/623453/71382765-36036f80-258e-11ea-94ea-3c13e57be2b2.png)

Can we please just add first class support in pilot so we don't need all of these hacky Envoyfilters are destroy performance?

@mandarjog @kyessenov

**Comment by kyessenov** [comment]
Just to confirm: there's no protobuf merge exercised for this envoy filter?

**Comment by howardjohn** [comment]
it's not merging but it has to unmarshal, clone, and remarshal. Cloning is also doing a merge internally. The config is already any marshalled which is why we don't just append to a list

**Comment by mandarjog** [comment]
Makes sense that any istio feature not cause regressions in pilot.

@kyessenov we still do think that the underlying wasm config api will evolve after 1.5, or will it not change that much?

I would fix the EnvoyFilter impl so that adding 2 filters does not incur such a price. There are customers inserting multiple lua filters today and are paying this cost. 

There will be a new istio extensions api to make this better.

@rshriram perhaps we need to update the EnvoyFilter processing so that we don’t end up multiple round trips and proto merges for the “insert this filter” use case?

**Comment by kyessenov** [comment]
Re: Wasm config API -- that depends on how upstreaming goes, and the timing of the sync-up to istio. The config API is very likely to undergo some change (naming bike-shed).

**Comment by howardjohn** [comment]
https://github.com/istio/istio/pull/19786 should help a lot, will run some additional tests to see how much exactly

**Comment by mandarjog** [comment]
Thanks @hzxuzhonghu . Please check if you can additionally optimize the common case of “I want to add these filters”

**Comment by howardjohn** [comment]
I don't see why the API not being stable impacts "native" implementation in pilot. We can just have pilot send different config for 1.4 proxies and 1.5 proxies, just like we do for many other features?

**Comment by kyessenov** [comment]
@howardjohn The issue is that we need something more generic than 1.4 vs 1.5. Ultimately, people can bring their own envoy, so we should permit changes to the config proto in the future without any code change. Where we can improve the performance, is in the selection mechanism since that is constrained by pilot implementation and is not easily upgradable. So the general shape of the extension API is:
```yaml
match: # OR of conditions
- <condition on the proxy (version, role, workload), protocol, port, service>
position: 
  before: HTTP_AUTH # before or after 
  # filter chain to inject a filter (these are merged in their order here): 
  # tcp.telemetry, tcp.attribute, tcp.auth, tcp.router, http.telemetry, http.attribute, http.auth, http.router, upstream.router  
filters: # Filters inserted together to respect relative order
- name: envoy.wasm
  config: {} # google.protobuf.Any with the config
```

**Comment by howardjohn** [comment]
I am not sure why we need to couple telemetry v2 to having a good mechanism for sending and configuring generic wasm filters for Envoy. The latter is much further off, is it not?

We can have telemetry v2 work perfectly today with minimal effort (correct me if I'm wrong here, but I would expect very minimal amount of pilot code) and then later transition to the extensions model once it's ready

**Comment by kyessenov** [comment]
@howardjohn While it might be faster to do in the short term, it's creating cruft for the future and might hurt long term. We already have too many one-off solutions for various extensions (this envoy filter, fault, mixer filter, authn/z, now telemetry v2?) that are difficult to upgrade and migrate between control plane implementations. We want something more principled than envoy filter, but not as hard-coded as say fault filter, so we just have to select the right middle ground.

**Comment by howardjohn** [comment]
I am not sure those are one-off, they are using the existing mechanism for extending Pilot - plugins. Realistically, how soon are we going to create this new API, implement it, stabilize it, move telemetry v2 over to wasm, and move telemetry v2 over to use this new API? 

Writing this in pilot is not hard to upgrade, that is a solved problem that impacts every feature. I don't think it's any worse than EnvoyFilter, but it removes the need to have a bunch of envoy filters for each proxy version.

If we are really trying to support other control planes as well it's much simpler to have them add a new plugin that just sends the wasm config than implement the entire EnvoyFilter spec (which is really not feasible unless they are outputting the same exact XDS config as we are which is unlikely).

I am not sure why you think a Pilot implementation is more crufty than EnvoyFilter. The API docs even say it is break-glass, yet we are making telemetry, a core Istio feature, configured through it?

What are our 1.5 plans? Will we make v2 default?

**Comment by kyessenov** [comment]
If we push this down to Pilot code, and given our release cycle, we'd need to push complexity into envoy to maintain compatibility with the existing extension and Wasm config. We certainly want to move fast there and not having to satisfy Pilot's expectations for now.

**Comment by howardjohn** [comment]
We have always had a requirement to upgrade pilot before the proxy. Pilot
can support multiple versions of the config, so I do not see how it
restricts velocity of breaking the API every few weeks

**Comment by kyessenov** [comment]
I'm thinking of slowly moving trains like GKE add-on. Once we ship 1.5, we're likely have to live with it for the next year. I do not think Wasm or telemetry v2 config will look the same in a year.

**Comment by howardjohn** [comment]
Are we optimizing for a case where the control planes is static yet the
data plane is upgraded? I am not sure that scenario exista? but maybe I am
not aware of future plans/use cases

**Comment by kyessenov** [comment]
Yes, exactly, that's the whole point of Wasm. We show be able to add features to old installations without any change to envoy or pilot.

**Comment by howardjohn** [comment]
Is there a design doc detailing how pilot will support that?

**Comment by kyessenov** [comment]
No, there's not but this new extension API design is in scope.

**Comment by howardjohn** [comment]
With the recent perf improvements, this isn't too bad anymore. It would be nice to have native performance, but I don't think this is P0 anymore

**Comment by mandarjog** [comment]
@howardjohn can you share measurements from the latest master? Thanks.

**Comment by howardjohn** [comment]
Looks like without envoyfilters, its around ~550m cpu and with its at ~700m. There were some other changes that optimized pilot further so I may need a bigger cluster to get more accurate numbers, since at such a low CPU the error margin is fairly high
